### ISSUE TITLE
Corrupted/garbage graphics after the Nov 4th wip build

### ISSUE THREAD

**Issue by dezraj** [gonetz/GLideN64#2161]
Hello gonetz! So this is what is happening after the Nov 4th wip build.

I'm having corrupted/garbage/artifact graphics as below. At first I thought it only happened with hires textures, but actually it is happening with all the games. Take a look:

![pj64gfx3](https://user-images.githubusercontent.com/27839354/71616434-07d8ed80-2b95-11ea-82a1-166315edb00f.jpg)
![pj64gfx4](https://user-images.githubusercontent.com/27839354/71616443-10c9bf00-2b95-11ea-8846-58cb4ecdb260.jpg)
![pj64gfx5](https://user-images.githubusercontent.com/27839354/71616452-1921fa00-2b95-11ea-8d88-0de513cd3ddb.jpg)
![pj64gfx6](https://user-images.githubusercontent.com/27839354/71616461-20490800-2b95-11ea-8203-11493d451664.jpg)

My setup:
Win 8.1 x64
i7 3517u
intel HD graphics 4000 + nvidia GTX 740m
Project64 2.3.3

The Nov 4th build works normally, so something changed after this build.

**Comment by gonetz** [comment]
I can't find such an old  hardware to check the problem.
Luckily, there are not so many code changes since  Nov 4th.
My guess is that the problems are caused by either by rewriting of blender code or by enabling threadedVideo option by default.

Since I have no such issues on my hardware, I need your help to find the problem commit.
First, disable threaded video in config dialog. Check, that the problem is still exists.

Then, I made several builds to check:
https://drive.google.com/file/d/13oCMSi1Zvsyeoy9ZUDWeMQyEp5QHggRN/view?usp=sharing

Please check, with which build the regression appears.

**Comment by dezraj** [comment]
Hello gonetz! Thank you for your support.

Threaded video was already disabled when I did the first tests, so I skipped this one and went straight to the test builds. Here are the results:

build 0: ok.
build 1: ok.
build 2: bad.
build 3: bad.
build 4: bad.
build 5: bad.
build 6: bad.

All the "bad" builds had threaded video enabled by default. Disabling or enabling this option had no change in the results. The same for builds 0 and 1, enabling or disabling that option didn't trigger the bug, so I suppose the problem is not related to this.

Thanks again for your time.

**Comment by gonetz** [comment]
Thanks! We are almost there. There are only 4 commits between build 1 and build 2 which can cause the problem. Now we need to find the problem commit. I made 4 new test builds:
https://drive.google.com/file/d/1AzS5AyXRdSL9LafpEy2y_DtZ90483y-O/view?usp=sharing

You need to find the first build with the problem. Most likely it will be build 1.1, but I need to be sure.

**Comment by dezraj** [comment]
Hello gonetz! Here are the results for test builds 2:

build 1.1: ok
build 1.2: ok
build 1.3: bad
build 1.4: bad

By the way, testing the Nov 4th build again I think it has a little better fps stability when playing the F Zero X demo, for example, compared to the good test builds. But I suppose this is not related to this issue.

Thank you so much again for your great support.

**Comment by gonetz** [comment]
> build 1.2: ok
> build 1.3: bad

Wow! It is surprise for me. So, commit a01718d2 introduced the problem.
It looks harmless. Have you seen some error reports in gliden64.log?

Anyway. here a new test build:
https://drive.google.com/file/d/1c0XRYXSiRUAsPN5WaYDfqs2SOOofbWGP/view?usp=sharing

I made a correction, which may fix the problem. I'm saying "may be" because I don't know what causes the problem, I can only guess. Please test it.

**Comment by dezraj** [comment]
Hey gonetz! I don't know where to find this file gliden64.log.

This new test build doesn't fix the problem unfortunately.

If you have any other ideas please let me know and I'll make other tests.

Thanks again.

**Comment by gonetz** [comment]
GLideN64 creates gliden64.log in the same folder where the plugin placed. It logs various errors, especially with shaders compilation. The issues you experiencing look as if some shaders compiled wrong. Please try to find it.

I also prepared another test build:
https://drive.google.com/file/d/14yL9CSsRapoipL7POJz9W_keWGCjqOql/view?usp=sharing

It is my last blind guess. I have no more ideas how to fix it.

**Comment by dezraj** [comment]
Hello gonetz! Thanks for your time again. Unfortunately this build didn't fix the problem either. I suppose I'll have to stick to the Nov 4th build for a while.

By the way, I can't see this gliden64.log file anywhere within Project64's folders. Only GLideN64.ini is generated. Perhaps this file not being created is an issue too?

Anyway, thank you for your support again.

**Comment by gonetz** [comment]
> Perhaps this file not being created is an issue too?

No, it rather means that shaders compiled without errors.

**Comment by fzurita** [comment]
@gonetz Try providing a build with link time optimization turned off.

**Comment by gonetz** [comment]
Ok, here a build with link time optimization turned off:
https://drive.google.com/file/d/1cqrWjvFN5E5L2mSK1TdKEbJoN-xzdYIL/view?usp=sharing
@dezraj please test it too.

**Comment by dezraj** [comment]
fzurita and gonetz, the bug persists unfortunately. :(

**Comment by fzurita** [comment]
Can you try deleting your shader cache?

**Comment by gonetz** [comment]
@fzurita The problem is caused by changing 

`lowp vec2 needMirror = step(vec2(0.5), mod(intPart, vec2(2.0))) * vMirror;`
by
`lowp vec2 needMirror = step(vec2(0.5), mod(intPart, vWrap)) * vMirror;`

This is hardly somehow related to plugin optimization or shaders cache.

You told me that default value of vWrap is too high and can cause incorrect work of mod.
We rewrote this shader so incorrect execution of mod does not affect the result.
Anyway, I set defaults for vWrap to 10.0 in my test builds, but it did not help.
@dezraj can't find gliden64.log, so shaders compilation completed without errors.

**Comment by fzurita** [comment]
Ok. I have seen strange issues that have been resolved by closing the shader cache.

This is a weird issue.

**Comment by dezraj** [comment]
Hello gonetz and fzurita! I cleared the shaders folder but the corrupted/garbage graphics persisted with the last provided test build. Thank you so much guys for your support. Tell me if you need me testing anything else. ;)

**Comment by dezraj** [comment]
Hey guys! I guess I've figured out what was causing the bug. I right-clicked my desktop and noticed nvidia control panel wasn't showing as an option in the context menu. So I searched online and found that this is caused by something wrong with the Intel integrated graphics driver. So I updated my Intel graphics driver and the nvidia control panel backed working again.

I then set my nvidia graphics as my preferred graphics card in the nvidia control panel.

Out of curiosity I went to test the latest wip build again and everything is normal now with the textures.

So I suppose we can close this issue (unless it was specifically related to intel integrated graphics, which I haven't tested, but I doubt it was, I'm inclined to think something must have gone wrong with the the way the hybrid graphics work on my laptop).

Thank you again for your time.
